### PR TITLE
ST-09001: Harden Core Tool Composition Contracts

### PR DESCRIPTION
**Pull request by TVScoundrel**
## Story
- ST-09001: Harden Core Tool Composition Contracts

## What Changed
- Refactored `packages/core/src/tools/composition.ts` to replace broad `any` signatures with generic `ComposedTool<TInput, TOutput>` and runtime-erased `unknown` boundaries.
- Added focused helper utilities (`isConditionalStep`, `calculateRetryDelay`, `toError`) to separate orchestration flow from utility logic.
- Added retry-option validation guard to fail fast when `maxAttempts < 1`.
- Fixed timeout-resource handling by clearing scheduled timeout handles after `Promise.race` settles.
- Added focused tests in `packages/core/tests/tools/composition.test.ts` covering sequential, parallel, conditional, compose, retry, timeout, and cache behavior.
- Added story documentation at `docs/st09001-core-tool-composition-contracts.md` including warning deltas and validation evidence.
- Updated planning trackers/checklists and moved ST-09001 to In Review.

## Acceptance Criteria Coverage
- [x] `packages/core/src/tools/composition.ts` replaces broad `any`-based public contracts with generic input/output types (`unknown` where runtime-erased)
- [x] Composition helpers are split into clearer typed boundaries where needed (SRP-focused extraction)
- [x] Behavior remains backward compatible for existing call sites in `core` tests
- [x] Focused tests are added/updated for composition behavior and type-sensitive edge cases
- [x] Explicit-`any` warning deltas in touched files are reduced and recorded in story documentation
- [x] Story documentation added at `docs/st09001-core-tool-composition-contracts.md`

## Validation Performed
- `pnpm exec eslint packages/core/src/tools/composition.ts packages/core/tests/tools/composition.test.ts`
- `pnpm test --run packages/core/tests/tools/composition.test.ts` (9 passed)
- `pnpm lint:explicit-any:baseline`
  - `packages/**/src/**/*.ts`: `385 -> 372`
  - `core`: `189 -> 176`
- `pnpm test --run`
  - `147 passed | 16 skipped` files
  - `2084 passed | 286 skipped` tests
- `pnpm lint`
  - exit `0`; warnings only (`0` errors)

## Status
- Ready for review.

### CODE REVIEW

File: packages/core/src/tools/composition.ts [hunk 1]
```diff
@@ -142,16 +183,21 @@ export function retry(
...
+  ms: number
+): ComposedTool<TInput, TOutput> {
   return {
     name: `timeout(${tool.name})`,
     description: `${tool.description} (with ${ms}ms timeout)`,
-    invoke: async (input: any) => {
+    invoke: async (input: TInput): Promise<TOutput> => {
+      const timeoutPromise = new Promise<never>((_, reject) =>
+        setTimeout(() => reject(new Error(`Tool ${tool.name} timed out after ${ms}ms`)), ms)
+      );
```
[the remark this answers is not in the document]
Comment (TVScoundrel) [reply packages/core/src/tools/composition.ts]: Addressed in fbf3c85: timeout now stores the handle and clears it in a finally block after Promise.race settles. Added a regression test that verifies no pending timer remains when the wrapped tool completes before the timeout.
